framework/Runners/ExternalRunner.py

```diff
@@ -1,274 +0,0 @@
-
-    if os.path.exists(executable):
-      executableFile = open(executable, "r")
-
-      firstTwoChars = executableFile.read(2)
-
-      if firstTwoChars == "#!":
-        realExecutable = shlex.split(executableFile.readline())
-        self.raiseAMessage("reading #! to find executable:" + repr(realExecutable))
-        # The below code should work, and would be better than findMsys,
-        # but it doesn't work.
-        # winExecutable = subprocess.check_output(['cygpath','-w',realExecutable[0]],shell=True).rstrip()
-        # print("winExecutable",winExecutable)
-        # realExecutable[0] = winExecutable
-        def findMsys():
-          """
-            Function to try and figure out where the MSYS64 is.
-            @ In, None
-            @ Out, dir, String, If not None, the directory where msys is.
-          """
-          dir = os.getcwd()
-          head, tail = os.path.split(dir)
-          while True:
-            if tail.lower().startswith("msys"):
-              return dir
-            dir = head
-            head, tail = os.path.split(dir)
-          return None
-        msysDir = findMsys()
-        if msysDir is not None:
-          beginExecutable = realExecutable[0]
-          if beginExecutable.startswith("/"):
-            beginExecutable = beginExecutable.lstrip("/")
-          winExecutable = os.path.join(msysDir, beginExecutable)
-          self.raiseAMessage("winExecutable" + winExecutable)
-          realExecutable[0] = winExecutable
-        else:
-          self.raiseAWarning("Could not find msys in "+os.getcwd())
-        commandSplit = realExecutable + [executable] + commandSplit[1:]
-        return commandSplit
-    return origCommand
-
-
-  def start(self):
-    """
-      Function to run the driven code
-      @ In, None
-      @ Out, None
-    """
-    oldDir = os.getcwd()
-    os.chdir(self.workingDir)
-    localenv = dict(os.environ)
-    outFile = open(self.output,'w', self.bufferSize)
-    if platform.system() == 'Windows':
-      self.command = self.expandForWindows(self.command)
-      self.raiseAMessage("modified command to" + repr(self.command))
-    self.__process = utils.pickleSafeSubprocessPopen(self.command,shell=True,stdout=outFile,stderr=outFile,cwd=self.workingDir,env=localenv)
-    os.chdir(oldDir)
-    self.started = True
-
-  def kill(self):
-    """
-      Function to kill the subprocess of the driven code
-      @ In, None
-      @ Out, None
-    """
-    self.raiseAWarning("Terminating "+self.__process.pid+' '+self.command)
-    self.__process.terminate()
-
-  def getWorkingDir(self):
-    """
-      Function to get the working directory path
-      @ In, None
-      @ Out, workingDir, string, working directory
-    """
-    return self.workingDir
-
-  def getOutputFilename(self):
-    """
-      Function to get the output filenames
-      @ In, None
-      @ Out, self.output, string, output filename root
-    """
-    return os.path.join(self.workingDir,self.output)
```

Comment:
Sorry, it got a bit long-winded. I hope it makes sense. If we can do it better, I wouldn't be opposed to another iteration of improvement to this redesign. Some of these things were cobbled together as I went, and on an as-needed basis. I may be lacking vision on some longer term aspects of RAVEN however.

I put a comment below in the DynamicEventTree that further discusses some of this that kind of describes one of the problems I ran into with having an ExternalRunner that sits next to the InternalRunner.  It may be that we can use the ExternalRunner in different contexts?  I don't know, if we think eventually something other than Codes will need to generate subprocesses, then it may be useful to keep revisit whether an ExternalRunner can be used.